tempesta_fw/http_parser.c

```diff
@@ -2430,274 +2444,310 @@ __req_parse_referer(TfwHttpMsg *hm, unsigned char *data, size_t len)
-			parser->_i_st = &&I_DateHour;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_MonthBeg) {
+		if ('A' <= c && c <= 'Z') {
+			parser->month_int =
+				((size_t)c) << 24 | (parser->month_int >> 8);
+			__NEXT_TEMPL_STATE();
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc * 3600;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateMinCln, __fsm_n);
+		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateMinCln) {
-		if (likely(c == ':'))
-			__FSM_I_MOVE(I_DateMin);
+	__FSM_STATE(I_Month) {
+		if ('a' <= c && c <= 'z') {
+			parser->month_int =
+				((size_t)c) << 24 | (parser->month_int >> 8);
+			__NEXT_TEMPL_STATE();
+		}
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateMin) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_a(p, __fsm_sz, colon_a, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateMin;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_Year) {
+		if ('0' <= c && c <= '9') {
+			parser->date.year = parser->date.year * 10 + (c - '0');
+			__NEXT_TEMPL_STATE();
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc * 60;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateSecCln, __fsm_n);
+		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateSecCln) {
-		if (likely(c == ':'))
-			__FSM_I_MOVE(I_DateSec);
+	__FSM_STATE(I_Hour) {
+		if ('0' <= c && c <= '9') {
+			parser->date.hour = parser->date.hour * 10 + (c - '0');
+			__NEXT_TEMPL_STATE();
+		}
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateSec) {
-		__fsm_sz = __data_remain(p);
-		__fsm_n = parse_int_ws(p, __fsm_sz, &parser->_acc);
-		if (__fsm_n == CSTR_POSTPONE) {
-			parser->_i_st = &&I_DateSec;
-			__msg_hdr_chunk_fixup(data, len);
+	__FSM_STATE(I_Min) {
+		if ('0' <= c && c <= '9') {
+			parser->date.min = parser->date.min * 10 + (c - '0');
+			__NEXT_TEMPL_STATE();
 		}
-		if (__fsm_n < 0)
-			return __fsm_n;
-		parser->_date += parser->_acc;
-		parser->_acc = 0;
-		__FSM_I_MOVE_n(I_DateSecSP, __fsm_n);
+		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateSecSP) {
-		if (likely(c == ' '))
-			__FSM_I_MOVE(I_DateZone);
+	__FSM_STATE(I_Sec) {
+		if ('0' <= c && c <= '9') {
+			parser->date.sec = parser->date.sec * 10 + (c - '0');
+			__NEXT_TEMPL_STATE();
+		}
 		return CSTR_NEQ;
 	}
 
-	__FSM_STATE(I_DateZone) {
-		TRY_STR("gmt", I_DateZone, I_EoL);
+	__FSM_STATE(I_GMT) {
+		TRY_STR_BY_REF("gmt", &&I_GMT,
+		               st[parser->date.type][parser->date.pos + 1]);
 		TRY_STR_INIT();
 		return CSTR_NEQ;
 	}
 
+	__FSM_STATE(I_Res) {
+		unsigned int month;
+		time_t date;
+
+		month = __parse_month(parser->month_int);
+		if (month < 0)
+			return CSTR_NEQ;
+
+		if (parser->date.year < 100)
+			parser->date.year += (parser->date.year < 70) ? 2000
+			                                              : 1900;
+
+		date = __date_secs(parser->date.year, month,
+		                   parser->date.day, parser->date.hour,
+		                   parser->date.min, parser->date.sec);
+		if (date < 0)
```

Comment:
Won't this make `Thu,  1 Jan 1970 00:00:00 GMT` invalid? Some client may use that date in `If-Modified-Since`, and that looks like a correct date to me.